Penalty history improvements

featheredtoast opened the pull request:
Two features here:

1: Add a history count when applying penalties to give moderators context to the user they are penalizing.

2: Add a site setting that designates default penalty values in hours.

Silence/suspend modals will auto-fill in the default values, but otherwise
will still allow moderators to pick and overwrite values as normal.

First silence/suspend: first value
Second silence/suspend: second value
etc.

Penalty counts are forgiven at the same rate as tl3 promotion requirements do.

![Screenshot from 2021-06-10 17-51-22](https://user-images.githubusercontent.com/1322534/121614868-81c10780-ca14-11eb-942b-396181decaff.png)

![Screenshot from 2021-06-10 17-51-48](https://user-images.githubusercontent.com/1322534/121614897-900f2380-ca14-11eb-9670-7c4c39880987.png)

![Screenshot from 2021-06-10 17-52-31](https://user-images.githubusercontent.com/1322534/121614945-a74e1100-ca14-11eb-8749-ca0b79727d85.png)

featheredtoast committed:
FEATURE: add penalty history when silencing a user

featheredtoast committed:
add penalty history display on suspend

featheredtoast committed:
FEATURE: allow default penalty values to be chosen

Adds a site setting that designates default penalty values in hours.

Silence/suspend modals will auto-fill in the default values, but otherwise
will still allow moderators to pick and overwrite values as normal.

First silence/suspend: first value
Second silence/suspend: second value
etc.

Penalty counts are forgiven at the same rate as tl3 promotion requirements do.

jjaffeux commented on app/assets/javascripts/admin/addon/controllers/modals/admin-silence-user.js:
```diff
@@ -12,6 +13,14 @@ export default Controller.extend(PenaltyController, {
     this.setProperties({ silenceUntil: null, silencing: false });
   },
 
+  @discourseComputed("user")
```
could you explain the purpose of this and why it's done this way please?

We can't keep it that way, but I want to better understand before I advise

featheredtoast replied:
I have a situation where I need an initial value from the user object (the user's next penalty) even though the value isn't loaded when the component is rendered. The user might need to load after the modal shows (https://github.com/featheredtoast/discourse/blob/0542e9cb19bc4c2bc2a39d56011e03c1742f3f95/app/assets/javascripts/admin/addon/services/admin-tools.js#L52-L68).

I cannot make silenceUntil/suspendUntil itself a computed property here because I need to allow the future-date-picker the ability to call mut on the value, uninhibited.

I need to update silenceUntil/suspendUntil's initial value when the user loads in to the modal, but I am unable to within the same runloop as the init (Ember complains that we're recomputing the value too many times). Putting it in a later() allows me to update the initial value quickly enough.

I *could* load the user first, and *then* display the modal with the initial values, but if the user needs to load and takes a long time, no loading indicator would be shown so I decided to go down this route.

I need the following events to fire:
* Modal window shows, value is potentially blank as user loads
* User loads in. Fires an event to the controller that sets the initial value of silenceUntil/suspendUntil from user object. This should also update the future-date-picker's selection to custom, with the specified date/time along with the underlying input value for said date. Without the later() call it does update the UI, but not the underlying js value as a `mut` call was never triggered from the future-date-picker.

featheredtoast replied:
@jjaffeux let me know if the above makes sense to you!

jjaffeux commented on app/assets/javascripts/admin/addon/controllers/modals/admin-suspend-user.js:
```diff
@@ -12,6 +13,14 @@ export default Controller.extend(PenaltyController, {
     this.setProperties({ suspendUntil: null, suspending: false });
   },
 
+  @discourseComputed("user")
```
same

jjaffeux committed:
avoids later

featheredtoast committed:
FIX: future-date-input only initialize date/time selection once

Only initialize date/time selection from initial input once.
Otherwise, this may re-render from the input again. May revert selection
of future date input.

featheredtoast committed:
Refactor: function to calculate user's next_penalty

featheredtoast committed:
UX: Add display polish to penalty counter

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/discourse/discourse?pullRequest=13359) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/discourse/discourse?pullRequest=13359) before we can accept your contribution.<br/>**18** out of **19** committers have signed the CLA.<br/><br/>:white_check_mark: tgxworld<br/>:white_check_mark: jjaffeux<br/>:white_check_mark: eviltrout<br/>:white_check_mark: markvanlan<br/>:white_check_mark: nbianca<br/>:white_check_mark: lis2<br/>:white_check_mark: pmusaraj<br/>:white_check_mark: ZogStriP<br/>:white_check_mark: udan11<br/>:white_check_mark: romanrizzi<br/>:white_check_mark: techAPJ<br/>:white_check_mark: xfalcox<br/>:white_check_mark: CvX<br/>:white_check_mark: AndrewPrigorshnev<br/>:white_check_mark: eltociear<br/>:white_check_mark: featheredtoast<br/>:white_check_mark: jbrw<br/>:white_check_mark: martin-brennan<br/>:x: dependabot[bot]<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/discourse/discourse?pullRequest=13359) it.</sub>

featheredtoast commented:
Updated UI suggestions, added UI indicators if user has previous suspensions:

![Screenshot from 2021-07-01 19-27-09](https://user-images.githubusercontent.com/1322534/124211719-75275080-daa2-11eb-825a-952b47b6c957.png)

Added @jjaffeux 's suggestions for removing the defer functions. I needed to re-implement the initial input setup in init() rather than on didReceiveAttrs, but other than that it worked flawlessly (Thank you!)

Added @jjaffeux 's suggestions for golfing user.next_penalty functions

featheredtoast committed:
REFACTOR: move penalty history counts into its own component

featheredtoast commented:
rebased and cleaned up history so it reads cleanly and it doesn't include every commit under the sun according to discoursebot
